streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -16,276 +16,98 @@
 
     private final TimeTrackerSupplier sharedTimeTrackerSupplier;
 
-    KStreamKStreamJoin(final boolean isLeftSide,
-                       final String otherWindowName,
-                       final JoinWindowsInternal windows,
-                       final ValueJoinerWithKey<? super K, ? super V1, ? super V2, ? extends VOut> joiner,
-                       final boolean outer,
-                       final Optional<String> outerJoinWindowName,
-                       final TimeTrackerSupplier sharedTimeTrackerSupplier) {
-        this.isLeftSide = isLeftSide;
+    KStreamKStreamJoin(final String otherWindowName,
+            final JoinWindowsInternal windows,
+            final long joinBeforeMs,
+            final long joinAfterMs,
+            final ValueJoinerWithKey<? super K, ? super V1, ? super V2, ? extends VOut> joiner,
+            final Optional<String> outerJoinWindowName,
+            final TimeTrackerSupplier sharedTimeTrackerSupplier,
+            final boolean outer) {
         this.otherWindowName = otherWindowName;
-        if (isLeftSide) {
-            this.joinBeforeMs = windows.beforeMs;
-            this.joinAfterMs = windows.afterMs;
-        } else {
-            this.joinBeforeMs = windows.afterMs;
-            this.joinAfterMs = windows.beforeMs;
-        }
         this.windowsAfterMs = windows.afterMs;
         this.windowsBeforeMs = windows.beforeMs;
-        this.joinGraceMs = windows.gracePeriodMs();
+        this.joinBeforeMs = joinBeforeMs;
+        this.joinAfterMs = joinAfterMs;
         this.enableSpuriousResultFix = windows.spuriousResultFixEnabled();
         this.joiner = joiner;
-        this.outer = outer;
+        this.joinGraceMs = windows.gracePeriodMs();
         this.outerJoinWindowName = outerJoinWindowName;
         this.sharedTimeTrackerSupplier = sharedTimeTrackerSupplier;
+        this.outer = outer;
     }
 
-    @Override
-    public Processor<K, V1, K, VOut> get() {
-        return new KStreamKStreamJoinProcessor();
-    }
-
-    private class KStreamKStreamJoinProcessor extends ContextualProcessor<K, V1, K, VOut> {
-        private WindowStore<K, V2> otherWindowStore;
-        private Sensor droppedRecordsSensor;
-        private Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>>> outerJoinStore = Optional.empty();
-        private InternalProcessorContext<K, VOut> internalProcessorContext;
-        private TimeTracker sharedTimeTracker;
+    protected abstract class KStreamKStreamJoinProcessor extends ContextualProcessor<K, V1, K, VOut> {
+        protected WindowStore<K, V2> otherWindowStore;
+        protected Sensor droppedRecordsSensor;
+        protected Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>>> leftOuterJoinStore =
+                Optional.empty();
+        protected Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V2, V1>>> rightOuterJoinStore =
+                Optional.empty();
```

Comment:
The left store is only used by the Left class, and the right store is only used by the Right class. These can be combined into a single outerJoinStore, and the KStreamKStreamJoinProcessor can take generic arguments <Va, Vb> that determine the types in the joinStore signature.

The LeftJoinProcessor can then set Va=V1 and Vb=V2, while the RightJoinProcessor sets Va=V2 and Vb=V1.